scheduler/rank_test.go

```diff
@@ -127,6 +127,246 @@ func TestBinPackIterator_NoExistingAlloc(t *testing.T) {
+
+	// Verify network information at task level
+	require.Equal(300, out[0].TaskResources["web"].Networks[0].MBits)
+	require.Equal(300, out[1].TaskResources["web"].Networks[0].MBits)
+}
+
+// Tests that bin packing iterator fails due to overprovisioning of network
+// This test has network resources at task group and task level
+func TestBinPackIterator_Network_Failure(t *testing.T) {
+	_, ctx := testContext(t)
+	nodes := []*RankedNode{
+		{
+			Node: &structs.Node{
+				// 50% fit
+				NodeResources: &structs.NodeResources{
+					Cpu: structs.NodeCpuResources{
+						CpuShares: 4096,
+					},
+					Memory: structs.NodeMemoryResources{
+						MemoryMB: 4096,
+					},
+					Networks: []*structs.NetworkResource{
+						{
+							Mode:   "host",
+							Device: "eth0",
+							CIDR:   "192.168.0.100/32",
+							MBits:  1000,
+						},
+					},
+				},
+				ReservedResources: &structs.NodeReservedResources{
+					Cpu: structs.NodeReservedCpuResources{
+						CpuShares: 1024,
+					},
+					Memory: structs.NodeReservedMemoryResources{
+						MemoryMB: 1024,
+					},
+					Networks: structs.NodeReservedNetworkResources{
+						ReservedHostPorts: "1000-2000",
+					},
+				},
+			},
+		},
+	}
+
+	// Add a planned alloc that takes up some network mbits at task and task group level
+	plan := ctx.Plan()
+	plan.NodeAllocation[nodes[0].Node.ID] = []*structs.Allocation{
+		{
+			AllocatedResources: &structs.AllocatedResources{
+				Tasks: map[string]*structs.AllocatedTaskResources{
+					"web": {
+						Cpu: structs.AllocatedCpuResources{
+							CpuShares: 2048,
+						},
+						Memory: structs.AllocatedMemoryResources{
+							MemoryMB: 2048,
+						},
+						Networks: []*structs.NetworkResource{
+							{
+								Device: "eth0",
+								IP:     "192.168.0.1",
+								MBits:  300,
+							},
+						},
+					},
+				},
+				Shared: structs.AllocatedSharedResources{
+					Networks: []*structs.NetworkResource{
+						{
+							Device: "eth0",
+							IP:     "192.168.0.1",
+							MBits:  400,
+						},
+					},
+				},
+			},
+		},
+	}
+	static := NewStaticRankIterator(ctx, nodes)
+
+	// Create a task group with networks specified at task and task group level
+	taskGroup := &structs.TaskGroup{
+		EphemeralDisk: &structs.EphemeralDisk{},
+		Tasks: []*structs.Task{
+			{
+				Name: "web",
+				Resources: &structs.Resources{
+					CPU:      1024,
+					MemoryMB: 1024,
+					Networks: []*structs.NetworkResource{
+						{
+							Device: "eth0",
+							MBits:  300,
+						},
+					},
+				},
+			},
+		},
+		Networks: []*structs.NetworkResource{
+			{
+				Device: "eth0",
+				MBits:  500,
+			},
+		},
+	}
+
+	binp := NewBinPackIterator(ctx, static, false, 0)
+	binp.SetTaskGroup(taskGroup)
+
+	scoreNorm := NewScoreNormalizationIterator(ctx, binp)
+
+	out := collectRanked(scoreNorm)
+	require := require.New(t)
+
+	// We expect a placement failure because we need 800 mbits of network
+	// and only 200 is free
```

Comment:
Might be worth explaining logic here.  I was expecting 300 Mbits to be available:  The node declares 1000 Mbits, and the planned alloc has 300+400=700 Mbits allocated - this leaves 300 Mbits left?

Also, might be useful to update values such the network bandwidth of task and shared is each less than available bandwidth but exceed bandwidth when combined.  In this example, we can set the taskGroup network specs Mbits to 170 for example.

Reply:
comment was wrong, will fix. I like your test suggestion, will change it to that condition.